### ISSUE TITLE
String-Compression Idea: Next-string index MISS encoding option

### ISSUE THREAD

**Issue by kannanvijayan-zz** [binast/binjs-ref#234]
I bounced some of our discussions against Iain just picked his brain on the string-ref compression issue.  He proposed a very interesting approach to further handling string window - and especially handling string-window misses.

His suggestion seems to neatly address two issues in a very clean way in one go:

1. The issue of string-window misses.
2. The issue of large JSON-style object literals with long sequences of unique keys.

The approach itself is as follows:

1. First, we order the table of strings on first-use.  This is a neutral choice, but it plays well with chunking and streaming compilation (you just add strings to the end of the string table as you see new unique strings).

2. Under that ordering during compress/decompress, if we see a new "MISS" there is an elevated probability that the newly seen string has never been seen before.  This is especially true if we are in the middle of one of those large JSON object literals

3. Using this intuition, we differentiate the "MISS" probability into two probabilities: a) the new string's index has never been seen before, in which case its index MUST be exactly one higher than the index of most recently encoded unique string (due to how we order our string table).  or b) the new string's index has already been seen once, in which case this is a "infrequently used" string.

These two cases can be labelled "MISS-NEXT" and "MISS-RANDOM", and exposed as individual symbols which can be encoded.

When encoding the "MISS-NEXT" symbol, no string-ref number needs to be encoded as the decoder can simply compute that.  If the probability tables for the string-window indexes (including the "MISS-NEXT" symbol) are contextualized on path, then the probabilities of "MISS-NEXT" will naturally be increased in "dictionary-style object literals" - where the path of  `DataProperty.key::StringLiteral.value` would select heavily for "MISS-NEXT" rather than one of the actual indices.

I'm highly optimistic about this approach addressing a good chunk of our string-window-miss issues.  Perhaps not all off them (in particular, the "MISS-RANDOM" strings might need additional work), but still a really strong win.

Pinging @Yoric and @dominiccooney 

Tried to find Iain's github profile, but failed.  I'll send him the link to this issue over IRC.

**Comment by kannanvijayan** [comment]
Ping @IainIreland

**Comment by iainireland** [comment]
Sorry, I'm still getting up to speed. What's the static prediction table? Why doesn't this work with it?

**Comment by iainireland** [comment]
One more idea that Kannan and I batted around over lunch:

The encoder can determine whether a string is going to be reused within the string window. (The obvious trivial case: if a string is unique, it will never be reused.) There's no point in putting strings into the window if they won't be reused; they just push out other strings and reduce the probability of reuse. In addition to MISS-NEXT and "MISS-RANDOM", we could also have "MISS-NEXT-UNIQUE", which means "take the next unused string in the string table, but don't add it to the window". Intuitively, it seems like this would be quite rare for identifiers, somewhat common for properties, and quite common for string literals.

(We could also consider other variations, like MISS-RANDOM-UNIQUE.)

**Comment by kannanvijayan** [comment]
> That wouldn't work for strings that are not part of the static prediction table, which is our most critical goal here.

I don't see how this would need a static prediction table.  On the compressor side, we do the following:

1. If window-hit, encode using window.
2. If window-miss, check the string-table index K+1 where K is the highest stringref index seen so far (i.e. the most recent unique string), if the string-ref index == K+1, then encode a MISS-NEXT
3. Otherwise encode a MISS-RANDOM followed by the directly-encoded stringref index.

On the decompressor, it's pretty straightforward:  if a MISS-NEXT shows up, just emit the string at index K+1 (K tracked symmetrically in the decoder as it is in the encoder).

**Comment by Yoric** [comment]
@kannanvijayan Oh, when you write `DataProperty.key::StringLiteral.value`, that's the path, not the value itself?

**Comment by kannanvijayan** [comment]
@Yoric yes - that was an example of the path-suffix context that would be present inside JSON literals.  Sorry for the confusing post.

**Comment by Yoric** [comment]
Ok, yeah, in that case, it should work. Move-to-front techniques are somewhat expensive for decompression, but it's worth trying.

**Comment by Yoric** [comment]
For the moment, I'm prioritizing having a full (with roundtrip) extensible version with brotli (#208 and more), which will serve us as a measure of progress. Once this is done, I'll return to evaluating window compression.

**Comment by Yoric** [comment]
So, I'm actually implementing some of this together with brotli at the moment:

- #301 
- #302 
- #304 

with pretty good results. Eventually, we should check whether we can achieve similarly good results with Entropy coding instead of Brotli compression.
